From 68e75402b12dafd74e5c22e297552095a6beb7fd Mon Sep 17 00:00:00 2001
From: Madhu <enometh@net.meer>
Date: Sun, 7 Jun 2020 11:28:57 +0530
Subject: [PATCH] extend ffi:defcallback to support exported symbols

* src/cmp/cmppass1-ffi.lsp: (c1-defcallback): extend the form which the
NAME argument to FFI:DEFCALLBACK can take: if NAME is a list of three
elements then a non-NIL third element indicates that the callback should
be exported.  In this case the c-name of the callback is derived from
NAME via LISP-TO-C-NAME.

* src/cmp/cmpbackend-cxx/cmppass2-top.lsp: (t3defcallback): arrange to
export this c-name with extern linkage

This patch extends ffi:defcallback so the NAME argument can be
of the form (lisp-name call-type export-p)

If EXPORT-P is non-NIL build-shared-library will arrange to export an
ELF symbol, derived via (c::lisp-to-c-name lisp-name), which can be
directly accessed from C.

e.g.  - with (c::install-c-compiler)

$cat hello.lisp
(ffi:defcallback (hello :default :export-p) :int () 42)

$ ecl --norc --eval '(compile-file "hello.lisp" :system-p t)' --eval '(ext:quit)'
$ ecl --norc --eval "(c:build-shared-library \"hello\" :lisp-files '(\"hello.o\") :init-name \"init_hello\")" --eval '(ext:quit)'
$ objdump -T libhello.so | grep hello

$cat hello_eg.c
int
main(int argc, char **argv)
{
	extern void init_hello(cl_object);
	extern int hello(void);
	cl_boot(argc, argv);
	ecl_init_module(NULL, init_hello);
	printf("%d\n", hello());
}

$ gcc hello_eg.c -L . -lhello -ldl -lecl
$ LD_LIBRARY_PATH=`pwd` ./a.out

240516 - t3callback
---
 src/cmp/cmpbackend-cxx/cmppass2-top.lsp | 14 +++++++++++---
 src/cmp/cmppass1-ffi.lsp                | 25 ++++++++++++++++++-------
 2 files changed, 29 insertions(+), 10 deletions(-)

diff --git a/src/cmp/cmpbackend-cxx/cmppass2-top.lsp b/src/cmp/cmpbackend-cxx/cmppass2-top.lsp
index 85fdcfe3d..8348e67de 100644
--- a/src/cmp/cmpbackend-cxx/cmppass2-top.lsp
+++ b/src/cmp/cmpbackend-cxx/cmppass2-top.lsp
@@ -265,7 +265,9 @@
     (wt-nl-close-many-braces 0)))
 
 (defun t3callback (lisp-name c-name c-name-constant return-type return-type-code
-                       arg-types arg-type-constants call-type &aux (return-p t))
+                       arg-types arg-type-constants call-type
+                       export-callback-p
+                       &aux (return-p t))
   (declare (ignore lisp-name))
   (with-bir-env (:env 0 :level 0 :volatile "volatile ")
     (when (eql return-type :void)
@@ -278,8 +280,14 @@
                   (:stdcall "__stdcall ")
                   (t (cmperr "DEFCALLBACK does not support ~A as calling convention"
                              call-type)))))
-      (wt-nl-h "static " return-type-name " " fmod c-name "(")
-      (wt-nl1  "static " return-type-name " " fmod c-name "(")
+      (wt-nl-h (if export-callback-p
+                   "extern "
+                   "static ")
+               return-type-name " " fmod c-name "(")
+      (wt-nl1  (if export-callback-p
+                   "__attribute__((visibility(\"default\"))) "
+                   "static ")
+               return-type-name " " fmod c-name "(")
       (loop with comma = ""
             for var in vars
             for type in arg-types
diff --git a/src/cmp/cmppass1-ffi.lsp b/src/cmp/cmppass1-ffi.lsp
index 84be2eeef..466fd3cc6 100644
--- a/src/cmp/cmppass1-ffi.lsp
+++ b/src/cmp/cmppass1-ffi.lsp
@@ -138,6 +138,13 @@
 ;;; types like :fixnum or :sse2. Another argument against such
 ;;; approach is semantic equivalence between interpreted and compiled
 ;;; versions of the special form. -- jd 2019-11-27
+
+;; ;m 200607 - extend the form which NAME can take: if name is a list
+;; of three elements then a non-NIL third element indicates that the
+;; callback should be exported.  In this case the c-name of the
+;; callback is derived from NAME via LISP-TO-C-NAME.  Arrange to export
+;; this c-name with extern linkage via t3-defcallback.
+
 (defun c1-defcallback (args)
   (destructuring-bind (name return-type arg-list &rest body)
       args
@@ -146,12 +153,15 @@
           ((and (consp return-type)
                 (member (first return-type) '(* array)))
            (setf return-type :pointer-void)))
-    (let ((arg-types '())
-          (arg-type-constants '())
-          (arg-variables '())
-          (c-name (format nil "ecl_callback_~d" (length *callbacks*)))
-          (name (if (consp name) (first name) name))
-          (call-type (if (consp name) (second name) :cdecl)))
+    (let* ((export-callback-p (and (consp name) (third name)))
+           (arg-types '())
+           (arg-type-constants '())
+           (arg-variables '())
+           (c-name (if export-callback-p
+                       (lisp-to-c-name (first name))
+                       (format nil "ecl_callback_~d" (length *callbacks*))))
+           (name (if (consp name) (first name) name))
+           (call-type (if (consp name) (second name) :cdecl)))
       (dolist (i arg-list)
         (unless (consp i)
           (cmperr "Syntax error in CALLBACK form: C type is missing in argument ~A "i))
@@ -164,7 +174,8 @@
                   (foreign-elt-type-code return-type)
                   (reverse arg-types)
                   (reverse arg-type-constants)
-                  call-type)
+                  call-type
+                  export-callback-p)
             *callbacks*)
       (c1expr
        `(progn
-- 
2.39.2.101.g768bb238c4

